### PR TITLE
[core] Updated change_element_subtype method

### PR DESCRIPTION
**Pull request by Zioba**
* [ ] Read PR [documentation](http://ostis-dev.github.io/sc-machine/dev/pr/)
* [ ] Update changelog
* [ ] Update documentation

Resolves: #381

### CODE REVIEW

**Comment by deniskoronchik** [comment]
- Make and link issue.
- Rename commit

File: sc-memory/sc-core/sc-store/sc_storage.c [hunk 1]
```diff
@@ -793,9 +793,6 @@ sc_result sc_storage_change_element_subtype(const sc_memory_context *ctx, sc_add
   sc_element *el = null_ptr;
   sc_result r = SC_RESULT_OK;
 
-  if (type & sc_type_element_mask)
-    return SC_RESULT_ERROR_INVALID_PARAMS;
-
   if (sc_storage_element_lock(addr, &el) != SC_RESULT_OK)
```
Comment (deniskoronchik) [inline sc-memory/sc-core/sc-store/sc_storage.c:796]: This breaks sc-memory consistency

File: sc-memory/sc-memory/sc_scs_helper.cpp [hunk 2]
```diff
@@ -169,8 +169,29 @@ class StructGenerator
...
+              {
+                  //Type NodeConst means that element have been called by system Id, that is why need to ignore this case as error
+                  if (newType != ScType::NodeConst)
+                  {
+                      SC_THROW_EXCEPTION(utils::ExceptionInvalidType, "Duplicate element type for " + el.GetIdtf());
+                  }
+              }
+          }
+      }
+
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_scs_helper.cpp:194]: tabulation 2 spaces

File: sc-memory/tests/scs/units/test_scs_level_6.cpp [hunk 3]
```diff
@@ -70,6 +70,26 @@ TEST(scs_level_6, smoke)
   }
 }
 
+TEST(scs_level_6, changing_type)
+{
+    std::string const data =
+            "a -> b;;"
+            "sc_node_struct -> a;;";
```
Comment (deniskoronchik) [inline sc-memory/tests/scs/units/test_scs_level_6.cpp:77]: Why this test in scs_level_6?

**Review by deniskoronchik — CHANGES_REQUESTED** [review]
Discuss

File: sc-memory/sc-memory/sc_scs_helper.cpp [hunk 4]
```diff
@@ -169,7 +169,26 @@ class StructGenerator
         {
           SetSCsGlobalIdtf(el.GetIdtf(), result);
         }
-        
+      }
+      else
+      {
+        ScType const &newType = el.GetType();
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_scs_helper.cpp:175]:
```suggestion
        ScType const & newType = el.GetType();
```

File: sc-memory/sc-memory/sc_scs_helper.cpp [hunk 5]
```diff
@@ -169,7 +169,26 @@ class StructGenerator
...
+        ScType const &oldType = m_ctx.GetElementType(result);
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_scs_helper.cpp:176]:
```suggestion
        ScType const & oldType = m_ctx.GetElementType(result);
```

File: sc-memory/sc-memory/sc_scs_helper.cpp [hunk 6]
```diff
@@ -169,7 +169,26 @@ class StructGenerator
...
+          {
+            m_ctx.SetElementSubtype(result, *newType);
+          }
+          else
+          {
+            //Type NodeConst means that element have been called by system Id, that is why need to ignore this case as error
+            if (newType != ScType::NodeConst)
+            {
+              SC_THROW_EXCEPTION(utils::ExceptionInvalidType, "Duplicate element type for " + el.GetIdtf());
+            }
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_scs_helper.cpp:189]:
```suggestion
              SC_THROW_EXCEPTION(utils::ExceptionInvalidType, "Element " + el.GetIdtf() + " already  defined as type " + .... + " before. New type is " + ....);
```

**Review by deniskoronchik — CHANGES_REQUESTED** [review]
Fixes

**Review by KovalM — APPROVED** [review]

**Review by deniskoronchik — CHANGES_REQUESTED** [review]
I don't understand these fixes at all

File: sc-memory/sc-core/sc-store/sc_storage.c (hunk 1 above)
Comment (Zioba) [reply sc-memory/sc-core/sc-store/sc_storage.c:799]: This does not break consistency, This is just check, that new type is existing type.
By the way, check doesn't work

File: sc-memory/sc-core/sc-store/sc_storage.c [hunk 7]
```diff
@@ -805,6 +802,25 @@ sc_result sc_storage_change_element_subtype(const sc_memory_context *ctx, sc_add
...
+  else if ((old_type & sc_type_arc_mask))
+  {
+    if (!(type & sc_type_arc_mask))
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+  else if ((old_type & sc_type_link))
+  {
+    if (!(type & sc_type_link))
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
```
Comment (KovalM) [inline sc-memory/sc-core/sc-store/sc_storage.c:822]:
```suggestion
  if (!((type & sc_type_element_mask) & (old_type & sc_type_element_mask)))
    return SC_RESULT_ERROR_INVALID_PARAMS;
```

**Review by KovalM — CHANGES_REQUESTED** [review]

**Review by deniskoronchik — CHANGES_REQUESTED** [review]
Proposal to close this PR at all. It brakes full logic of API

File: sc-memory/sc-memory/sc_memory.hpp [hunk 8]
```diff
@@ -107,7 +107,7 @@ class ScMemoryContext
   //! Returns type of sc-element. If there are any error, then returns ScType::Unknown
   _SC_EXTERN ScType GetElementType(ScAddr const & addr) const;
 
-  /*! Change subtype of sc-element (subtype & sc_type_element_mask == 0).
+  /*! Change subtype of sc-element ((type & sc_type_element_mask) & (old_type & sc_type_element_mask) == 0).
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_memory.hpp:110]:
```suggestion
  /*! Change subtype of sc-element
```

File: sc-memory/sc-core/sc-store/sc_storage.h [hunk 9]
```diff
@@ -92,10 +92,11 @@ sc_addr sc_storage_arc_new_ext(sc_memory_context * ctx, sc_type type, sc_addr be
...
 sc_result sc_storage_get_element_type(sc_memory_context const * ctx, sc_addr addr, sc_type * result);
 
-/*! Change element sub-type
- * @param addr sc-addr of element to set new type
- * @param type New sub-type of sc-element (this type must be: type & sc_type_element_mask == 0)
- * @return If sub-type changed, then returns SC_RESULT_OK; otherwise returns SC_RESULT_ERROR
+/*! Change element subtype
+ * @param addr sc-addr of element to set new subtype
+ * @param type New type of sc-element (this basic type must be the same as type of existing element:
+ * (type & sc_type_element_mask) & (old_type & sc_type_element_mask) == 0)
```
Comment (deniskoronchik) [inline sc-memory/sc-core/sc-store/sc_storage.h:98]:
```suggestion
 * @param type New type of sc-element
```

File: sc-memory/sc-core/sc-store/sc_storage.c [hunk 10]
```diff
@@ -805,6 +802,15 @@ sc_result sc_storage_change_element_subtype(const sc_memory_context *ctx, sc_add
...
 
+  sc_type old_type = el->flags.type;
+
+  if (!(old_type & sc_type_element_mask))
+    return SC_RESULT_ERROR_INVALID_STATE;
+
+  //Sc-memory consistency protection
+  if (!((type & sc_type_element_mask) & (old_type & sc_type_element_mask)))
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+
```
Comment (deniskoronchik) [inline sc-memory/sc-core/sc-store/sc_storage.c:813]: if ((el->flags.type & sc_type_element_mask) != (type & sc_type_element_mask))
  return SC_RESULT_ERROR_INVALID_PARAMS;

File: sc-memory/sc-core/sc-store/sc_storage.c [hunk 11]
```diff
@@ -805,6 +802,15 @@ sc_result sc_storage_change_element_subtype(const sc_memory_context *ctx, sc_add
...
   if (sc_access_lvl_check_write(ctx->access_levels, el->flags.access_levels))
     el->flags.type = (el->flags.type & sc_type_element_mask) | (type & ~sc_type_element_mask);
```
Comment (deniskoronchik) [inline sc-memory/sc-core/sc-store/sc_storage.c:815]:
```suggestion
    el->flags.type = type;
```

File: sc-memory/sc-memory/sc_scs_helper.cpp [hunk 12]
```diff
@@ -169,7 +169,26 @@ class StructGenerator
...
           SetSCsGlobalIdtf(el.GetIdtf(), result);
         }
-        
+      }
+      else
+      {
+        ScType const & newType = el.GetType();
+        ScType const & oldType = m_ctx.GetElementType(result);
+        if (newType != oldType)
+        {
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_scs_helper.cpp:178]: Remove this if.

File: sc-memory/sc-memory/sc_scs_helper.cpp [hunk 13]
```diff
@@ -169,7 +169,26 @@ class StructGenerator
...
+          if (oldType.CanExtendTo(newType))
+          {
+            m_ctx.SetElementSubtype(result, *newType);
+          }
+          else
+          {
+            //Type NodeConst means that element have been called by system Id, that is why need to ignore this case as error
+            if (newType != ScType::NodeConst)
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_scs_helper.cpp:186]: Remove this if and comment

**Review by deniskoronchik — CHANGES_REQUESTED** [review]
Add tests for scs_helper to check logic.
- element created and type changed
- type change error
- invalid type assignment (edge type for node)
- etc.

File: sc-memory/sc-memory/sc_scs_helper.cpp (hunk 13 above)
Comment (Zioba) [reply sc-memory/sc-memory/sc_scs_helper.cpp:186]: if we remove this condition, the builder will not work

**Comment by Zioba** [comment]
> Add tests for scs_helper to check logic.
> 
> * element created and type changed
> * type change error
> * invalid type assignment (edge type for node)
> * etc.

These tests have been implemented and added in TEST_F(ScMemoryTest, elements) section 
in the file sc-memory/tests/sc-memory/common/sc-elements.cpp.

The only one existing test for changeSubtype method was here, so I placed new tests there.

**Comment by Zioba** [comment]
> Add tests for scs_helper to check logic.
> 
> * element created and type changed
> * type change error
> * invalid type assignment (edge type for node)
> * etc.

1) Added tests for scs-helper logic. At this level, only 1 feature has been added to the logic and must be tested: 
change type of node is possible only if the old type can be extended to a new type. 
For example: We can't change struct to tuple.

All other features are on the sc-memory level. That is why tests for them were implemented in sc-memory/tests/sc-memory/common/sc-elements.cpp

File: sc-memory/sc-memory/sc_scs_helper.cpp (hunk 12 above)
Comment (Zioba) [reply sc-memory/sc-memory/sc_scs_helper.cpp:178]: I'm not sure that this action speed up the work of the builder. Will check.

**Review by KovalM — APPROVED** [review]

File: sc-memory/sc-memory/sc_scs_helper.cpp (hunk 12 above)
Comment (Zioba) [reply sc-memory/sc-memory/sc_scs_helper.cpp:178]: done

File: sc-memory/sc-memory/sc_scs_helper.cpp [hunk 14]
```diff
@@ -169,7 +169,23 @@ class StructGenerator
...
+        }
+        else
+        {
+          //Type NodeConst means that element have been called by system Id, that is why need to ignore this case as error
+          if (newType != ScType::NodeConst)
+          {
+            SC_THROW_EXCEPTION(utils::ExceptionInvalidType, "Duplicate element type for " + el.GetIdtf());
+          }
+        }
       }
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_scs_helper.cpp:189]:
```cpp
if (!m_ctx.SetElementSubtype(result, *newType))
{
   SC_THROW_EXCEPTION(utils::ExceptionInvalidType, "Duplicate element type for " + el.GetIdtf());
}
```
Also as i mentioned before we should provide more error information. Which was type, which type we want to setup

File: sc-memory/tests/sc-memory/scs/scs-helper.cpp [hunk 15]
```diff
@@ -76,6 +76,42 @@ TEST_F(SCsHelperTest, GenerateBySCs_FileURL)
...
+  ScAddr node = m_ctx->CreateNode(ScType::Node);
+  m_ctx->HelperSetSystemIdtf("a", node);
+
+  EXPECT_TRUE(helper.GenerateBySCsText(data1));
+
+  ScType const nodeType = m_ctx->GetElementType(node);
+  EXPECT_EQ(nodeType, ScType::NodeConstStruct);
+
+  EXPECT_FALSE(helper.GenerateBySCsText(data2));
+  EXPECT_EQ(nodeType, ScType::NodeConstStruct);
```
Comment (deniskoronchik) [inline sc-memory/tests/sc-memory/scs/scs-helper.cpp:95]:
```suggestion
  std::string const data1 = "sc_node_struct -> a;;";
  std::string const data2 = "sc_node_tuple -> a;;";

  SCsHelper helper(*m_ctx, std::make_shared<TestFileInterface>());

  EXPECT_TRUE(helper.GenerateBySCsText(data1));

  ScAddr const node = m_ctx->HelperFindBySystemIdtf("a");
  EXPECT_TRUE(node);
  
  ScType const nodeType = m_ctx->GetElementType(node);
  EXPECT_EQ(nodeType, ScType::NodeConstStruct);

  EXPECT_FALSE(helper.GenerateBySCsText(data2));
  EXPECT_EQ(nodeType, ScType::NodeConstStruct);
```

File: sc-memory/tests/sc-memory/scs/scs-helper.cpp [hunk 16]
```diff
@@ -76,6 +76,42 @@ TEST_F(SCsHelperTest, GenerateBySCs_FileURL)
...
+}
+
+TEST_F(SCsHelperTest, GenerateBySCs_UsingExistingNonBasicElement)
```
Comment (deniskoronchik) [inline sc-memory/tests/sc-memory/scs/scs-helper.cpp:98]: Test that test nothing. It fully unusefull in this context.
Where tests that:
- try to change type of `node` to `link`?
- try to change structure type of one element twice in one scs?
- try to setup incorrect type to one element in scs:
  - ```sc_node_struct -> [];;```
  - ```sc_node_material -> [* x -> y;; *];;```
  - ```sc_node_material -> {};;```

**Review by deniskoronchik — CHANGES_REQUESTED** [review]
- Test coverage
- Better error message

**Review by deniskoronchik — APPROVED** [review]

**Review by KovalM — APPROVED** [review]

**Review by MikhailSadovsky — APPROVED** [review]

**Pull request merged**